### ISSUE TITLE
CTE does implicit and unexpected cross join when referencing columns that aren't in the table in the from clause

### ISSUE THREAD

**Issue by fisher-liquid** [duckdb/duckdb#15416]
### What happens?

When creating a CTE you are allowed to reference columns that are not in the table in the from clause of the CTE, and instead of getting an error it does a cross join (I think) which blows up the rows processed internally and the query time.

### To Reproduce

```
CREATE TABLE a as (
    SELECT row_number() OVER () as id1, row_number() OVER () as id2, random() as c1, random() as c2
    FROM generate_series(1,10000)
);
CREATE TABLE b as (
    SELECT row_number() OVER () as id1, row_number() OVER () as id2
    FROM generate_series(1,10000)
);
WITH b_cte as (
    -- table b does not have columns c1 or c2! why does this work
    SELECT id1, id2, c1, c2
    FROM b
)
SELECT a.id1, a.id2, a.c1, a.c2
FROM a
INNER JOIN b_cte ON b_cte.id1 = a.id1 and b_cte.id2 = a.id2;
```

See inside the CTE the columns `c1` and `c2` and not in table `b`, yet there is no error. EXPLAIN shows a projection up to many more rows than necessary (due to cross join between table b and a?):

```
┌───────────────────────────┐
│         PROJECTION        │
│    ────────────────────   │
│            id1            │
│            id2            │
│             c1            │
│             c2            │
│                           │
│       ~50000000 Rows      │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│      RIGHT_DELIM_JOIN     │
│    ────────────────────   │
│      Join Type: INNER     │
│                           │
│        Conditions:        │
│         id1 = id1         ├──────────────┐
│         id2 = id2         │              │
│ c1 IS NOT DISTINCT FROM c1│              │
│ c2 IS NOT DISTINCT FROM c2│              │
│                           │              │
│       ~50000000 Rows      │              │
└─────────────┬─────────────┘              │
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│         SEQ_SCAN          ││         HASH_JOIN         │
│    ────────────────────   ││    ────────────────────   │
│             a             ││      Join Type: INNER     │
│                           ││                           │
│        Projections:       ││        Conditions:        │
│             c1            ││         id1 = id1         ├───────────────────────────────────────────┐
│             c2            ││         id2 = id2         │                                           │
│            id1            ││ c1 IS NOT DISTINCT FROM c1│                                           │
│            id2            ││ c2 IS NOT DISTINCT FROM c2│                                           │
│                           ││                           │                                           │
│        ~10000 Rows        ││       ~50000000 Rows      │                                           │
└───────────────────────────┘└─────────────┬─────────────┘                                           │
                             ┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐
                             │       CROSS_PRODUCT       ├──────────────┐              │         DUMMY_SCAN        │
                             └─────────────┬─────────────┘              │              └───────────────────────────┘
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │         SEQ_SCAN          ││         DELIM_SCAN        │
                             │    ────────────────────   ││    ────────────────────   │
                             │             b             ││       Delim Index: 1      │
                             │                           ││                           │
                             │        Projections:       ││                           │
                             │            id1            ││                           │
                             │            id2            ││                           │
                             │                           ││                           │
                             │        ~10000 Rows        ││         ~5000 Rows        │
                             └───────────────────────────┘└───────────────────────────┘
```

### OS:

x86_64, Amazon Linux 2023 distro on ec2 instance

### DuckDB Version:

1.1.3

### DuckDB Client:

CLI and python

### Full Name:

Fisher Moritzburke

### Affiliation:

Liquid Analytics

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by kryonix** [comment]
This is technically expected behavior and not a bug. As your CTE `b_cte` is not `MATERIALIZED`, DuckDB inlines the definition. Hence the query that actually gets executed looks like this:

```sql
SELECT a.id1, a.id2, a.c1, a.c2
FROM a LATERAL -- ⚠ implicit!
INNER JOIN(SELECT id1, id2, c1, c2 FROM b) ON b_cte.id1 = a.id1 and b_cte.id2 = a.id2;
```

I've added the implicit `LATERAL` annotation. Now your previous CTE is actually a _correlated subquery_ and subject to query decorrelation. That's the reason you see the `RIGHT_DELIM_JOIN` and `DELIM_SCAN` in the plan.

**Comment by soerenwolfers** [comment]
MWE: 

```sql
WITH cte AS (
    SELECT x
)
SELECT *
FROM 
    (SELECT 1) _(x), 
    cte
;
```

@kryonix The behavior may be expected given the CTE is currently literally inlined, but should it really do that? To me, having variables leak like this feels spaghetti. The documentation also doesn't mention anywhere that `AS MATERIALIZED` vs `AS NOT MATERIALIZED` can have semantic consequences. Maybe could there be a `LATERAL` keyword on CTEs that explicitly allows this?

**Comment by soerenwolfers** [comment]
PostgreSQL doesn't allow this (since PostgreSQL requires an explicit `LATERAL` keyword, I need to make it a bit more complicated):

```sql
WITH cte AS (
    SELECT x 
)
SELECT cte FROM (SELECT 1) _(x), LATERAL (SELECT * FROM cte) cte
```
```
Query Error: column "x" does not exist
```

It doesn't even allow the following:

```sql
WITH cte AS (SELECT x)
SELECT 1;
```
```
Query Error: column "x" does not exist
```

**Comment by rpbouman** [comment]
Hi all,

> This is technically expected behavior and not a bug. As your CTE `b_cte` is not `MATERIALIZED`, DuckDB inlines the definition. Hence the query that actually gets executed looks like this:

I thought a couple of days about this explanation, and I think I now know why it doesn't sit well with me.
If one would explicitly mark a cte as MATERIALIZED, then I think the intent and semantics are clear: presumably the query author knows in advance that the cost of executing the query multiple times exceeds the cost of executing it once and caching and reusing its results in any subsequent references. 

But from this, it does not follow that when the cte is not marked as MATERIALIZED, it must mean that it is "inlined". Leaving out an explicit hint to materialize simply means the query author trusts DuckDB to manage query optimization. It does not mean the query author is telling DuckDB to "inline" the cte - whatever it means exactly. 

Considering the original example, apparently "inlining" literally means quite literally substituting the cte reference with the cte query definition - even if that means the cte can generate different results, depending upon the context or scope where it happens to get "inlined".

I don't think the behavior where the result of the query can be different depending upon whether the cte was "inlined" or materialized is desirable at all, and I don't believe other sql implementations behave that way. I can understand other people feel differently but in that case I would suggest to let query authors choose for this behavior by using explicit syntax. (For example, with an INLINED keyword where they can now write MATERIALIZED).

**Comment by kryonix** [comment]
Let's backtrack a bit here. This discussion went into an unexpected direction. Thank you all for thinking about this. Let me clarify a bit.

 _Of course_ the initial query here is bogus. And of course Postgres rejects it (rightfully so). But that was not my point. With the current way DuckDB handles CTEs and inlining, the behavior and the "unexpected" cross join _is actually expected_---that's all I was saying. Hence the _"technically"_ in my initial response. Whether that behavior is sound and should stay like it is, is a whole different question and discussion. I completely agree, that inlining in this case propably should not be done, as it is not intuitive, it leaks context etc. So, yes, there maybe should be a `column does not exist` error. But again, considering the current state of the system, it is _technically_ expected behavior.

---

> The documentation also doesn't mention anywhere that AS MATERIALIZED vs AS NOT MATERIALIZED can have semantic consequences.

There can be semantic consequences. _E.g._ if one materializes a query containing a volatile function, say `random`, it has a different semantic than not materializing it. It can also have effects in combination with recursive CTEs. Those cases may be very specific, but still, materialization can make semantic difference.

> Maybe could there be a LATERAL keyword on CTEs that explicitly allows this?

I don't think that this is really useful. See my point below regarding the `INLINED` keyword. I'd rather forbid the current behavior and refer users to `CREATE MACRO`-statements to achieve this.

---

> If one explicitly marks a cte as MATERIALIZED, then I think the intent and semantics are clear: presumably the query author knows in advance that the cost of executing the query multiple times exceeds the cost of executing it once and caching and reusing its results in any subsequent references. Similarly, marking the cte as NOT MATERIALIZED probably means the query author knows the cost of re-executing the query is probably lower than hanging on to the result and reusing that.

I completely agree with that.

> But from this, it does not follow that when the cte is not marked at all, it must mean that it is "inlined". Leaving out an explicit hint to materialize simply means the query author trusts DuckDB to manage query optimization. It does not mean the query author is telling DuckDB to "inline" the cte - whatever it means exactly.

I also agree with that. Automatic CTE materialization is an optimization DuckDB uses IFF (1) applicable, and (2) when the CTE is not annotated with any materialization keywords. However, the default behavior is inlining.

> Even when the cte is marked as NOT MATERIALIZED, it suggests simply that the result is not to be cached and reused - not that it must mean it should be "inlined".

We don't have that. A CTE is either materialized or inlined, similar to Postgres' capabilities. There is no third semantic (currently? maybe this discussion spawns one, who knows).

> I don't think the behavior where the result of the query can be different depending upon whether the cte was "inlined" or materialized is desirable at all, and I don't believe other sql implementations behave that way.

Other DBMSs absolutely behave similarly (if you can get them to inline your CTE 😉). Obviously not in the way of the initial query here, that's true. But for volatile functions, for example, inlining vs. materialization absolutly makes a difference. Apart from that I agree that materialization _mostly_ is an optimization that does not change semantics. In some cases it can, however.

> If duckdb can optimize by "inlining" a cte [...]

Inlining is not considered an optimization currently, but _the_ way how DuckDB handles CTEs.

> I can understand other people feel differently but in that case I would suggest to let query authors choose for this behavior by using explicit syntax. (For example, with an INLINED keyword where they can now write MATERIALIZED).

I do not think that any additional keywords are useful or required. What should be the difference between a `NOT MATERIALIZED` and `INLINED` CTE? Given that there are only two semantics, I think that's equivalent. Should inlining behave a bit less like macro expansion, and have a few more checks? Propably yes. I will look into this.

**Comment by rpbouman** [comment]
Hi @kryonix , thanks for the elaborate and nuanced reply - I'm learning a ton from it!

> ` I completely agree, that inlining in this case propably should not be done, as it is not intuitive, it leaks context etc`

This was the main point for me and I think this answers that completely to satisfaction. Thanks for that!

I will clear up a few item below, but this is just FYI - it is not my intent to complicate things.

> What should be the difference between a NOT MATERIALIZED and INLINED CTE? Given that there are only two semantics, I think that's equivalent.

Perhaps it is naive of me to think like this, but I imagined the engine executing the cte and then using the result immediately and only once where it is needed to resolve a reference to that cte, and that repeated usages of the cte would trigger another execution. 

Thanks to your explanation, I now understand that building the result of the cte in isolation is already materialization, and that this imaginary scenario would simply be materialization but without the benefit of waiting only once (or conversely, without the cost of keeping the result around for the remainder of the query).

**Comment by soerenwolfers** [comment]
> With the current way DuckDB handles CTEs and inlining, the behavior and the "unexpected" cross join is actually expected

I think "expected behavior" in the context of issue trackers means "not a bug; intended behavior". Otherwise, _every_ behavior is "expected behavior" in view of the current implementation. 

> There can be semantic consequences. E.g. if one materializes a query containing a volatile function, say random, it has a different semantic than not materializing it. It can also have effects in combination with recursive CTEs. Those cases may be very specific, but still, materialization can make a semantic difference.

I think that too is a bug. In PostgreSQL, 

```sql
WITH cte AS NOT MATERIALIZED (
    SELECT random()
)
SELECT 
    *
FROM
    cte a, cte b
```

returns the same value twice; in DuckDB, it returns two different values. In fact, [the documentation](https://www.postgresql.org/docs/current/queries-with.html) states that 

> if a WITH query is non-recursive and side-effect-free (that is, it is a SELECT containing no volatile functions) then it can be folded into the parent query, allowing joint optimization of the two query levels.

with an implied "if it is recursive or has side-effects (that is, it is _not_ a SELECT containing no volatile functions) then it cannot be folded into the parent query. 

Happy to open a separate issue about that if you think there is a chance this may be changed.

**Comment by kryonix** [comment]
Materialized CTEs were introduced precisely to fix such queries (https://github.com/duckdb/duckdb/pull/7126). It simply was not possible to express Postgres' behavior before. The heuristic to _not_ inline, when the CTE contains volatile functions is IMO sane, and should be applied. Achieving that is not so simple, however.

First, we have to change the default behavior of a system to view CTEs as materialized. Which is not trivially done without breaking a bunch of stuff. Then, inlining must be implemented as an optimization, rather than materialization being the optimization. There may be more to that 🤔 

Don't get me wrong. I'm absolutely not opposed to this. Quite the contrary. I have been thinking about this for a while now. I just have not found the time to do it properly.

**Comment by kryonix** [comment]
Regarding changing the default behavior of the system (which would fix the issue on hand), there was a related discussion here: https://github.com/duckdb/duckdb/issues/12875
